packages/hash/subgraph/src/stdlib/edge/link.ts

```diff
@@ -98,44 +98,45 @@ export const getIncomingLinksForEntityAtMoment = (
   const timestampString =
     typeof timestamp === "string" ? timestamp : timestamp.toISOString();
 
+  const entityEdges = subgraph.edges[entityId];
+
+  if (!entityEdges) {
+    return [];
+  }
+
   const uniqueEntitiesFilter = getUniqueEntitiesFilter();
 
-  /**
-   * @todo - replace this with the analogue of the code in `getOutgoingLinksForEntityAtMoment` when
-   *   reverse mappings are stored in the graph
-   *   https://app.asana.com/0/1201095311341924/1203399511264512/f
-   */
-  return Object.entries(subgraph.edges)
-    .filter(([_, inner]) => {
-      return Object.values(inner)
-        .flat()
-        .find(
-          (outwardEdge: OutwardEdge) =>
-            isHasRightEntityEdge(outwardEdge) &&
-            outwardEdge.rightEndpoint.baseId === entityId,
+  return (
+    Object.entries(entityEdges)
+      // Only look at edges that were created before or at the timestamp
+      .filter(([edgeTimestamp, _]) => edgeTimestamp <= timestampString)
+      // Extract the link `EntityEditionId`s from the endpoints of the link edges
+      .flatMap(([_, outwardEdges]) => {
+        return outwardEdges.filter(isIncomingLinkEdge).map((edge) => {
+          return edge.rightEndpoint;
+        });
+      })
+      .map(({ baseId: linkEntityId, timestamp: _firstEditionTimestamp }) => {
+        const linkEntity = mustBeDefined(
```

Comment:
It may be useful to have a message for the `mustBeDefined` here, it is a little cryptic when the throw message doesn't contain anything